File: results/dnd-rest-benchmark/lifecycle-runs/20260804T072004Z_claude_claude-sonnet-5_ruby-sinatra/routes/play/settlements.rb
# DM-managed campaign settlements: validated services and availability,
# discovered per player character.

VALID_SETTLEMENT_AVAILABILITY = %w[open limited closed].freeze

def play_settlement_payload(settlement, viewer_character_id: nil)
  discovered_by = JSON.parse(settlement['discovered_by_json'])
  discovered_by = discovered_by.include?(viewer_character_id) ? [viewer_character_id] : [] if viewer_character_id

  {
    settlement_id: settlement['settlement_id'],
    name: settlement['name'],
    services: JSON.parse(settlement['services_json']),
    availability: settlement['availability'],
    discovered_by: discovered_by
  }
end

# Validates name/services/availability shared by create and replace, and
# returns the normalized [name, services, availability] tuple.
def validate_settlement_fields!(body)
  name = body['name']
  services = body['services']
  availability = body['availability']

  halt 400, { error: 'invalid name' }.to_json unless name.is_a?(String) && !name.empty?
  halt 400, { error: 'invalid services' }.to_json unless services.is_a?(Array) && !services.empty?

  normalized = services.map { |s| s.is_a?(String) ? s.strip : nil }
  halt 400, { error: 'invalid services' }.to_json unless normalized.all? { |s| s && !s.empty? }
  halt 400, { error: 'invalid services' }.to_json unless normalized.uniq.length == normalized.length

  halt 400, { error: 'invalid availability' }.to_json unless VALID_SETTLEMENT_AVAILABILITY.include?(availability)

  [name, normalized, availability]
end

def find_play_settlement!(campaign_id, settlement_id)
  settlement = db.execute(
    'SELECT * FROM play_settlements WHERE campaign_id = ? AND settlement_id = ?',
    [campaign_id, settlement_id]
  ).first
  halt 404, { error: 'settlement not found' }.to_json unless settlement
  settlement
end

post '/v1/play/campaigns/:id/settlements' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_owner!(campaign, user)

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  settlement_id = body['settlement_id']
  halt 400, { error: 'invalid settlement_id' }.to_json unless settlement_id.is_a?(String) && !settlement_id.empty?

  name, services, availability = validate_settlement_fields!(body)

  existing = db.execute(
    'SELECT 1 FROM play_settlements WHERE campaign_id = ? AND settlement_id = ?',
    [campaign['id'], settlement_id]
  ).first
  halt 409, { error: 'settlement already exists' }.to_json if existing

  sequence = db.execute(
    'SELECT COALESCE(MAX(sequence), -1) + 1 AS seq FROM play_settlements WHERE campaign_id = ?',
    [campaign['id']]
  ).first['seq']

  db.execute(
    'INSERT INTO play_settlements (campaign_id, sequence, settlement_id, name, services_json, availability, discovered_by_json) VALUES (?, ?, ?, ?, ?, ?, ?)',
    [campaign['id'], sequence, settlement_id, name, services.to_json, availability, '[]']
  )

  status 201
  play_settlement_payload(find_play_settlement!(campaign['id'], settlement_id)).to_json
end

put '/v1/play/campaigns/:id/settlements/:settlement_id' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_owner!(campaign, user)

  settlement = find_play_settlement!(campaign['id'], params[:settlement_id])

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  name, services, availability = validate_settlement_fields!(body)

  db.execute(
    'UPDATE play_settlements SET name = ?, services_json = ?, availability = ? WHERE campaign_id = ? AND settlement_id = ?',
    [name, services.to_json, availability, campaign['id'], settlement['settlement_id']]
  )

  play_settlement_payload(find_play_settlement!(campaign['id'], settlement['settlement_id'])).to_json
end

post '/v1/play/campaigns/:id/settlements/:settlement_id/discover' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  is_owner = require_play_participant!(campaign, user, 'not a campaign member')
  halt 403, { error: 'forbidden' }.to_json if is_owner

  settlement = find_play_settlement!(campaign['id'], params[:settlement_id])

  member = db.execute(
    'SELECT character_id FROM play_campaign_members WHERE campaign_id = ? AND username = ?',
    [campaign['id'], user['username']]
  ).first
  character_id = member['character_id']

  discovered_by = JSON.parse(settlement['discovered_by_json'])

  if discovered_by.include?(character_id)
    status 200
  else
    discovered_by << character_id
    db.execute(
      'UPDATE play_settlements SET discovered_by_json = ? WHERE campaign_id = ? AND settlement_id = ?',
      [discovered_by.to_json, campaign['id'], settlement['settlement_id']]
    )
    status 201
  end

  settlement = find_play_settlement!(campaign['id'], settlement['settlement_id'])
  play_settlement_payload(settlement, viewer_character_id: character_id).to_json
end

get '/v1/play/campaigns/:id/settlements' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  is_owner = require_play_participant!(campaign, user, 'not a campaign member')

  settlements = db.execute(
    'SELECT * FROM play_settlements WHERE campaign_id = ? ORDER BY sequence ASC',
    [campaign['id']]
  )

  if is_owner
    { settlements: settlements.map { |s| play_settlement_payload(s) } }.to_json
  else
    member = db.execute(
      'SELECT character_id FROM play_campaign_members WHERE campaign_id = ? AND username = ?',
      [campaign['id'], user['username']]
    ).first
    own_character_id = member && member['character_id']

    visible = settlements.select { |s| JSON.parse(s['discovered_by_json']).include?(own_character_id) }
    { settlements: visible.map { |s| play_settlement_payload(s, viewer_character_id: own_character_id) } }.to_json
  end
end

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T002352Z_pi_kimi-k2p7-code_python-django/dndsite/views/settlements.py
"""Settlement views for play campaigns."""

import json
import sqlite3

from django.http import JsonResponse
from django.views.decorators.csrf import csrf_exempt

from .common import (
    require_actor,
    require_play_campaign_member_or_owner,
)
from ..http import bad_request, conflict, forbidden, not_found, parse_json, require_method
from ..db import db_conn


VALID_AVAILABILITY = {"open", "limited", "closed"}


def _normalize_services(services):
    """Validate and normalize a list of service strings.

    Returns the normalized list or raises ValueError on invalid input.
    """
    if not isinstance(services, list) or not services:
        raise ValueError
    seen = set()
    normalized = []
    for service in services:
        if not isinstance(service, str):
            raise ValueError
        trimmed = service.strip()
        if not trimmed:
            raise ValueError
        if trimmed in seen:
            raise ValueError
        seen.add(trimmed)
        normalized.append(trimmed)
    return normalized


def _settlement_record(row, discovered_by):
    """Build a settlement response dict from a row and discoverer list."""
    return {
        "settlement_id": row["settlement_id"],
        "name": row["name"],
        "services": json.loads(row["services_json"]),
        "availability": row["availability"],
        "discovered_by": discovered_by,
    }


def _dm_settlement_response(conn, row):
    """Return the full settlement response for the DM."""
    rows = conn.execute(
        "SELECT character_id FROM play_settlement_discoveries "
        "WHERE campaign_id = ? AND settlement_id = ? ORDER BY id ASC",
        (row["campaign_id"], row["settlement_id"]),
    ).fetchall()
    discovered_by = [r["character_id"] for r in rows]
    return _settlement_record(row, discovered_by)


def _player_settlement_response(conn, row, character_id):
    """Return the player-filtered settlement response."""
    discovered = conn.execute(
        "SELECT 1 FROM play_settlement_discoveries "
        "WHERE campaign_id = ? AND settlement_id = ? AND character_id = ?",
        (row["campaign_id"], row["settlement_id"], character_id),
    ).fetchone() is not None
    discovered_by = [character_id] if discovered else []
    return _settlement_record(row, discovered_by)


@csrf_exempt
def settlements(request, id):
    """GET /v1/play/campaigns/{id}/settlements or POST to create one."""
    bad = require_method(request, "GET", "POST")
    if bad:
        return bad
    if request.method == "POST":
        return _create_settlement(request, id)
    return _list_settlements(request, id)


def _create_settlement(request, id):
    actor, err = require_actor(request, "dm")
    if err is not None:
        return err

    body = parse_json(request)
    if body is None:
        return bad_request("invalid request")

    try:
        settlement_id = body["settlement_id"]
        name = body["name"]
        services = body["services"]
        availability = body["availability"]
        if not isinstance(settlement_id, str) or not settlement_id:
            raise ValueError
        if not isinstance(name, str) or not name:
            raise ValueError
        if not isinstance(availability, str) or availability not in VALID_AVAILABILITY:
            raise ValueError
        normalized_services = _normalize_services(services)
    except (KeyError, TypeError, ValueError):
        return bad_request("invalid request")

    with db_conn() as conn:
        campaign = conn.execute(
            "SELECT id, owner FROM play_campaigns WHERE id = ?", (id,)
        ).fetchone()
        if campaign is None:
            return not_found("campaign not found")
        if actor["username"] != campaign["owner"]:
            return forbidden()

        next_sequence = conn.execute(
            "SELECT COALESCE(MAX(sequence), 0) + 1 AS next_sequence "
            "FROM play_settlements WHERE campaign_id = ?",
            (id,),
        ).fetchone()["next_sequence"]

        try:
            conn.execute(
                "INSERT INTO play_settlements (campaign_id, settlement_id, name, services_json, availability, sequence) "
                "VALUES (?, ?, ?, ?, ?, ?)",
                (id, settlement_id, name, json.dumps(normalized_services), availability, next_sequence),
            )
        except sqlite3.IntegrityError:
            return conflict("settlement already exists")

        row = conn.execute(
            "SELECT campaign_id, settlement_id, name, services_json, availability FROM play_settlements "
            "WHERE campaign_id = ? AND settlement_id = ?",
            (id, settlement_id),
        ).fetchone()
        response = _dm_settlement_response(conn, row)

    return JsonResponse(response, status=201)


@csrf_exempt
def update_settlement(request, id, settlement_id):
    """PUT /v1/play/campaigns/{id}/settlements/{settlement_id}"""
    bad = require_method(request, "PUT")
    if bad:
        return bad

    actor, err = require_actor(request, "dm")
    if err is not None:
        return err

    body = parse_json(request)
    if body is None:
        return bad_request("invalid request")

    try:
        name = body["name"]
        services = body["services"]
        availability = body["availability"]
        if not isinstance(name, str) or not name:
            raise ValueError
        if not isinstance(availability, str) or availability not in VALID_AVAILABILITY:
            raise ValueError
        normalized_services = _normalize_services(services)
    except (KeyError, TypeError, ValueError):
        return bad_request("invalid request")

    with db_conn() as conn:
        campaign = conn.execute(
            "SELECT id, owner FROM play_campaigns WHERE id = ?", (id,)
        ).fetchone()
        if campaign is None:
            return not_found("campaign not found")
        if actor["username"] != campaign["owner"]:
            return forbidden()

        row = conn.execute(
            "SELECT campaign_id, settlement_id, name, services_json, availability FROM play_settlements "
            "WHERE campaign_id = ? AND settlement_id = ?",
            (id, settlement_id),
        ).fetchone()
        if row is None:
            return not_found("settlement not found")

        conn.execute(
            "UPDATE play_settlements SET name = ?, services_json = ?, availability = ? "
            "WHERE campaign_id = ? AND settlement_id = ?",
            (name, json.dumps(normalized_services), availability, id, settlement_id),
        )

        row = conn.execute(
            "SELECT campaign_id, settlement_id, name, services_json, availability FROM play_settlements "
            "WHERE campaign_id = ? AND settlement_id = ?",
            (id, settlement_id),
        ).fetchone()
        response = _dm_settlement_response(conn, row)

    return JsonResponse(response)


@csrf_exempt
def discover_settlement(request, id, settlement_id):
    """POST /v1/play/campaigns/{id}/settlements/{settlement_id}/discover"""
    bad = require_method(request, "POST")
    if bad:
        return bad

    actor, err = require_actor(request, "player")
    if err is not None:
        return err

    with db_conn() as conn:
        campaign = conn.execute(
            "SELECT id, owner FROM play_campaigns WHERE id = ?", (id,)
        ).fetchone()
        if campaign is None:
            return not_found("campaign not found")

        if actor["username"] == campaign["owner"]:
            return forbidden()

        member = conn.execute(
            "SELECT character_id FROM play_campaign_members "
            "WHERE campaign_id = ? AND username = ?",
            (id, actor["username"]),
        ).fetchone()
        if member is None:
            return forbidden()
        character_id = member["character_id"]

        row = conn.execute(
            "SELECT campaign_id, settlement_id, name, services_json, availability FROM play_settlements "
            "WHERE campaign_id = ? AND settlement_id = ?",
            (id, settlement_id),
        ).fetchone()
        if row is None:
            return not_found("settlement not found")

        existing = conn.execute(
            "SELECT 1 FROM play_settlement_discoveries "
            "WHERE campaign_id = ? AND settlement_id = ? AND character_id = ?",
            (id, settlement_id, character_id),
        ).fetchone()

        if existing is None:
            conn.execute(
                "INSERT INTO play_settlement_discoveries (campaign_id, settlement_id, character_id) "
                "VALUES (?, ?, ?)",
                (id, settlement_id, character_id),
            )
            status = 201
        else:
            status = 200

        response = _player_settlement_response(conn, row, character_id)

    return JsonResponse(response, status=status)


def _list_settlements(request, id):
    actor, campaign, err = require_play_campaign_member_or_owner(request, id)
    if err is not None:
        return err

    is_owner = actor["username"] == campaign["owner"]

    with db_conn() as conn:
        rows = conn.execute(
            "SELECT campaign_id, settlement_id, name, services_json, availability FROM play_settlements "
            "WHERE campaign_id = ? ORDER BY sequence ASC",
            (id,),
        ).fetchall()

        if is_owner:
            settlements = [_dm_settlement_response(conn, row) for row in rows]
        else:
            member = conn.execute(
                "SELECT character_id FROM play_campaign_members "
                "WHERE campaign_id = ? AND username = ?",
                (id, actor["username"]),
            ).fetchone()
            character_id = member["character_id"]
            settlements = []
            for row in rows:
                discovered = conn.execute(
                    "SELECT 1 FROM play_settlement_discoveries "
                    "WHERE campaign_id = ? AND settlement_id = ? AND character_id = ?",
                    (id, row["settlement_id"], character_id),
                ).fetchone() is not None
                if discovered:
                    settlements.append(_player_settlement_response(conn, row, character_id))

    return JsonResponse({"settlements": settlements})
